core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java

```diff
@@ -0,0 +1,297 @@
+            if (objectName == null) {
+                builder.startObject();
+            } else {
+                builder.startObject(objectName);
+            }
+            builder.field("index", command.shardId().index().name());
+            builder.field("shard", command.shardId().id());
+            builder.field("node", command.node());
+            builder.field("allocation_id", command.allocationId());
+            builder.field("ignore_deciders", command.ignoreDeciders());
+            builder.endObject();
+        }
+    }
+
+    private final ShardId shardId;
+    private final String node;
+    private final String allocationId;
+    private final boolean ignoreDeciders;
+
+    /**
+     * Create a new {@link AllocateAllocationCommand}
+     *
+     * @param shardId        {@link ShardId} of the shard to assign
+     * @param node           Node to assign the shard to
+     * @param allocationId   Allocation ID of the shard copy to assign the shard to
+     * @param ignoreDeciders should the shard be allowed to allocate independent of allocation deciders
+     */
+    public AllocatePrimaryAllocationCommand(ShardId shardId, String node, String allocationId, boolean ignoreDeciders) {
+        if (node == null && allocationId == null) {
+            throw new IllegalArgumentException("node and allocation id cannot be both null");
+        }
+        this.shardId = shardId;
+        this.node = node;
+        this.allocationId = allocationId;
+        this.ignoreDeciders = ignoreDeciders;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    /**
+     * Get the shards id
+     *
+     * @return id of the shard
+     */
+    public ShardId shardId() {
+        return this.shardId;
+    }
+
+    /**
+     * Get the id of the Node
+     *
+     * @return id of the Node
+     */
+    public String node() {
+        return this.node;
+    }
+
+    /**
+     * Get the allocation id
+     *
+     * @return allocation id
+     */
+    public String allocationId() {
+        return this.allocationId;
+    }
+
+    /**
+     * Determine if allocation deciders should be ignored
+     *
+     * @return <code>true</code> if allocation deciders are to be ignored. Otherwise <code>false</code>
+     */
+    public boolean ignoreDeciders() {
+        return this.ignoreDeciders;
+    }
+
+    @Override
+    public RerouteExplanation execute(RoutingAllocation allocation, ShardsAllocators shardsAllocators, boolean explain) {
+        DiscoveryNode discoNode = null;
+        if (node != null) {
+            discoNode = allocation.nodes().resolveNode(node);
+        }
+        final RoutingNodes routingNodes = allocation.routingNodes();
+
+        // check if we have a primary unassigned shard
+        ShardRouting shardRouting = null;
+        for (ShardRouting routing : routingNodes.unassigned()) {
+            if (routing.shardId().equals(shardId)) {
```

Comment:
Smoosh these into one if statement? Its not a requirement, I'm more curious why you didn't do it.